src/transformers/modeling_reformer.py

```diff
@@ -0,0 +1,1757 @@
+
+        query_key_dots = torch.matmul(query_vectors, key_vectors.transpose(-1, -2))
+
+        # free memory
+        del query_vectors, key_vectors
+
+        mask = self._compute_attn_mask(query_indices, key_indices, attention_mask, query_key_dots.shape)
+
+        if mask is not None:
+            # get mask tensor depending on half precision or not
+            if query_key_dots.dtype == torch.float16:
+                mask_value = self.mask_value_float16
+            else:
+                mask_value = self.mask_value_float32
+
+            query_key_dots = torch.where(mask, query_key_dots, mask_value)
+
+        # free memory
+        del mask
+
+        # softmax
+        logits = torch.logsumexp(query_key_dots, dim=-1, keepdim=True)
+        attention_probs = torch.exp(query_key_dots - logits)
+
+        # free memory
+        del logits
+
+        # dropout
+        attention_probs = nn.functional.dropout(attention_probs, self.dropout, self.training)
+
+        # Mask heads if we want to
+        if head_mask is not None:
+            attention_probs = attention_probs * head_mask
+
+        # attend values
+        out_vectors = torch.matmul(attention_probs, value_vectors)
+
+        # free memory
+        del value_vectors
+
+        # merge chunk length
+        out_vectors = self._merge_seq_length_dims(out_vectors, self.num_attention_heads)
+
+        assert out_vectors.shape == (batch_size, self.num_attention_heads, sequence_length, self.attention_head_size,)
+
+        out_vectors = self._merge_hidden_size_dims(out_vectors, self.num_attention_heads, self.attention_head_size)
+
+        if do_output_attentions is False:
+            attention_probs = ()
+
+        return LocalSelfAttentionOutput(hidden_states=out_vectors, attention_probs=attention_probs)
+
+    def _compute_attn_mask(self, query_indices, key_indices, attention_mask, query_key_dots_shape):
+        mask = None
+
+        # chunk attention mask and look before and after
+        if attention_mask is not None:
+            attention_mask = attention_mask.to(torch.uint8)[:, None, :]
+            attention_mask = self._split_seq_length_dim_to(attention_mask, -1, self.chunk_length, 1)
+            attention_mask_key = self._look_adjacent(attention_mask, self.num_chunks_before, self.num_chunks_after)
+
+        # Causal mask
+        if self.is_decoder is True:
+            mask = torch.ge(query_indices.unsqueeze(-1), key_indices.unsqueeze(-2)).to(query_indices.device)
+
+        # Attention mask
+        if attention_mask is not None:
+            # create attn_mask
+            attn_mask = (attention_mask.unsqueeze(-1) * attention_mask_key.unsqueeze(-2)).expand(query_key_dots_shape)
+            # multiply by casaul mask if necessary
+            if mask is not None:
+                mask = mask * attn_mask
+            else:
+                mask = attn_mask
+        return mask
+
+
+class ReformerSelfOutput(nn.Module):
```

Comment:
I have trouble with these little pairs of Linear + Dropout with fancy names, because I have no idea where the dropout is when reading the calling modules.

Reply:
I think this is equiv to
```python
nn.ModuleList([Linear(all_head_size, hid_size), nn.Dropout(hidden_dropout_prob)])
```